logging: Add support to switch logging format at runtime.

aasthagr opened the pull request:
Adding Support to Switch Logging Formats at Runtime with the consideration of making the design scalable to be able to add any additional logging formats in future with minimal overhead.

peter-mitsis commented on include/logging/log_backend.h:
```diff
@@ -254,6 +254,15 @@ static inline const struct log_backend *log_backend_get(uint32_t idx)
+ *
+ * @param[in] backend_name Name of the backend as defined by the LOG_BACKEND_DEFINE.
+ *
+ * @return Pointer to the backend.
```
Pointer to the backend if found, NULL if not found.

aasthagr replied:
Addressed.

peter-mitsis commented on include/logging/log_core.h:
```diff
@@ -895,6 +896,31 @@ do {									       \
 
 #include <syscalls/log_core.h>
 
+/** @brief Function pointer table for log output formats, indexed by enum log_format_flags.
```
Outdated comment; log_format_flags does not exist

aasthagr replied:
Addressed.

peter-mitsis commented on include/logging/log_core.h:
```diff
@@ -895,6 +896,31 @@ do {									       \
+
+/** @brief Set logging format.
+ *
+ * @param backend_name Backend Name Example: log_backend_uart.
```
Name of the backend. Example: "log_backend_uart"

(The period helps avoid potential ambiguity.)

aasthagr replied:
Modified the doxygen. Thanks!!

peter-mitsis commented on include/logging/log_output.h:
```diff
@@ -50,6 +50,12 @@ extern "C" {
  */
 #define LOG_OUTPUT_FLAG_FORMAT_SYST		BIT(7)
 
+#define LOG_OUTPUT_TEXT 0
+
+#define LOG_OUTPUT_SYST 1
+
+#define LOG_OUTPUT_DICT 2
+
```
Recommend adding a comment to indicate where/how these are used.

For example: /* Supported backend logging types for use with log_format_set() */

aasthagr replied:
Addressed.

peter-mitsis commented on include/logging/log_output.h:
```diff
@@ -121,6 +127,18 @@ void log_output_msg_process(const struct log_output *output,
 
+/** @brief Process log messages v2 to SYS-T format.
+ *
+ * Function is using provided context with the buffer and output function to
+ * process formatted string and output the data in sys-t format
```
This routine processes and outputs the supplied message in the SYS-T logging output format.

peter-mitsis commented on subsys/logging/log_core.c:
```diff
@@ -352,6 +359,43 @@ void log_n(const char *str,
+		return -EINVAL;
+	}
+
+	if (backend != NULL) {
+		if (backend->api->format_set == NULL) {
+			return -ENOTSUP;
+		} else {
+			return backend->api->format_set(backend, log_type);
+		}
+	} else {
+		return -EINVAL;
+	}
```
Simplify. If we follow a pattern of bailing out early, we can reduce indentation (and this helps keep it easier to read).

if (backend == NULL) {
    return -EINVAL;
}

if (backend->api->format_set == NULL) {
    return -ENOTSUP;
}

return backend->api->format_set(backend, log_type);

aasthagr replied:
Made the changes to simplify this.

peter-mitsis commented on subsys/logging/log_core.c:
```diff
@@ -352,6 +359,43 @@ void log_n(const char *str,
+
+	for (int i = 0; i < log_backend_count_get(); i++) {
+		backend = log_backend_get(i);
+		if (log_backend_is_active(backend)) {
```
Two items.
1. Not sure why we are saying that the backend needs to be active.
2. If it is truly necessary, then the API name should reflect that.

aasthagr replied:
Changed the name of the API to reflect the setting of log format for all active backends.

aasthagr replied:
> Unless I have missed it, the work appears to be incomplete. For example, I think the routine log_output_msg2_process() still has the following check. if (IS_ENABLED(CONFIG_LOG_MIPI_SYST_ENABLE) && flags & LOG_OUTPUT_FLAG_FORMAT_SYST) { log_output_msg2_syst_process(output, msg, flags); return; } There may be other locations that need to be cleaned up as well.

@peter-mitsis Have not modified the log_output_msg2_syst_process routine for now , since other backends except UART are still using the old implementation. Wanted to seek some reviews/approvals on the design part and following that I will make changes to log_output_msg2_syst_process routine along with other backends.

peter-mitsis commented on subsys/logging/log_core.c:
```diff
@@ -352,6 +359,43 @@ void log_n(const char *str,
+		if (backend->api->format_set == NULL) {
+			return -ENOTSUP;
+		} else {
+			return backend->api->format_set(backend, log_type);
```
The field "format_set" does not yet exist. It is currently added in the next commit.

aasthagr replied:
Added the declaration of format_set in the same commit now.

peter-mitsis commented on subsys/logging/Kconfig.backends:
```diff
@@ -56,6 +56,12 @@ config LOG_BACKEND_UART_OUTPUT_DICTIONARY_BIN
 
 endchoice
 
+config LOG_BACKEND_UART_OUTPUT_DEFAULT
```
I would recommend adding a comment here indicating that the numbering for LOG_BACKEND_UART_OUTPUT_TEXT,  .... be consistent across all backends in the Kconfig AND that their values must match the LOG_OUTPUT_xxx type in the C files.

aasthagr replied:
Added the comment indicating the details.

peter-mitsis commented on subsys/logging/Kconfig.backends:
```diff
@@ -56,6 +56,12 @@ config LOG_BACKEND_UART_OUTPUT_DICTIONARY_BIN
+config LOG_BACKEND_UART_OUTPUT_DEFAULT
+	int
+	default 0 if LOG_BACKEND_UART_OUTPUT_TEXT
+	default 1 if LOG_BACKEND_UART_SYST_ENABLE
```
The name LOG_BACKEND_UART_SYST_ENABLE is inconsistent with the others. I suggest changing this to follow the LOG_BACKEND_UART_OUTPUT_xxx style.

aasthagr replied:
Changed the LOG_BACKEND_UART_SYST_ENABLE to LOG_BACKEND_UART_OUTPUT_SYST. Will do the same for other backends when will add support for them after initial reviews on the design. Thanks

aasthagr commented:
@nordic-krch Ping. Could you please review this PR?

peter-mitsis reviewed (commented):
Unless I have missed it, the work appears to be incomplete. For example, I think the routine log_output_msg2_process() still has the following check.
        if (IS_ENABLED(CONFIG_LOG_MIPI_SYST_ENABLE) &&
            flags & LOG_OUTPUT_FLAG_FORMAT_SYST) {
                log_output_msg2_syst_process(output, msg, flags);
                return;
        }
There may be other locations that need to be cleaned up as well.

aasthagr commented:
> Unless I have missed it, the work appears to be incomplete. For example, I think the routine log_output_msg2_process() still has the following check. if (IS_ENABLED(CONFIG_LOG_MIPI_SYST_ENABLE) && flags & LOG_OUTPUT_FLAG_FORMAT_SYST) { log_output_msg2_syst_process(output, msg, flags); return; } There may be other locations that need to be cleaned up as well.

@peter-mitsis Have not modified the log_output_msg2_syst_process routine for now , since other backends except UART are still using the old implementation. Wanted to seek some reviews/approvals on the design part and following that I will make changes to log_output_msg2_syst_process routine along with other backends.

dcpleung commented on include/logging/log_core.h:
```diff
@@ -895,6 +896,31 @@ do {									       \
+ * @param log_type Log format.
+ *
+ * @retval -ENOTSUP If the backend does not support changing format types.
+ *         -EINVAL  If the input is invalid.
+ *			0 for success.
```
If you use `@retval`, you will need to use it for all individual return values.

aasthagr replied:
Addressed.

dcpleung commented on include/logging/log_core.h:
```diff
@@ -895,6 +896,31 @@ do {									       \
+ *
+ * @param log_type Log format.
+ *
+ * @retval Pointer to the last backend that failed.
+ *		   NULL for success.
```
(as mentioned above)

aasthagr replied:
Addressed.

dcpleung commented on include/logging/log_output.h:
```diff
@@ -121,6 +130,18 @@ void log_output_msg_process(const struct log_output *output,
+ * @param msg Log message.
+ * @param flag Optional flags.
+ */
+extern void log_output_msg2_syst_process(const struct log_output *log_output,
```
There is no need for `extern` here.

aasthagr replied:
Makes sense. Removed the extern keyword. Thanks for pointing it out.

MaureenHelm commented:
> Have not modified the log_output_msg2_syst_process routine for now , since other backends except UART are still using the old implementation. Wanted to seek some reviews/approvals on the design part and following that I will make changes to log_output_msg2_syst_process routine along with other backends.

I understand this argument for non-UART backends, but I would like to see at least the UART backend free of output format implementation details (i.e., remove references to `CONFIG_LOG_BACKEND_UART_OUTPUT_SYST`/`CONFIG_LOG_BACKEND_UART_SYST_ENABLE`) as part of this design concept.

aasthagr commented:
> CONFIG_LOG_BACKEND_UART_OUTPUT_SYST

@MaureenHelm The UART backend implementation for v2 logging subsystem does not use CONFIG_LOG_BACKEND_UART_OUTPUT_SYST/CONFIG_LOG_BACKEND_UART_SYST_ENABLE. It is only v1 logging subsystem using these configuration options. In this PR, I have just addressed the design for v2 logging, since from what I know v1 logging is being deprecated. Let me know your thoughts.

nordic-krch commented on subsys/logging/log_backend_uart.c:
```diff
@@ -103,14 +104,15 @@ static void process(const struct log_backend *const backend,
 	uint32_t flags = log_backend_std_get_flags();
 
-	flags |= IS_ENABLED(CONFIG_LOG_BACKEND_UART_SYST_ENABLE) ? LOG_OUTPUT_FLAG_FORMAT_SYST : 0;
+	void (*log_output_func)(const struct log_output*, struct log_msg2*,
```
you probably wanted to add it to log_msg2 process function, this one is deprecated. When you extend log_output api with getter function you can use it here.

aasthagr replied:
@nordic-krch  Did not get it. What is deprecated? If the function pointer table is called from log_output_msg2_process() function, then there would be no use of the table since the table selects the routine (log_output_msg2_process/ log_output_msg2_syst_process etc) that needs to be called based on Kconfig value for formats. Not using "flags" for setting of log format type as per this implementation.
Also, when you say log_output api, do you refer to log_output structure or log_output.c?

nordic-krch replied:
sorry, i got confused and did not notice the gap in the file so i thought that changes are in `put` function.

aasthagr replied:
No problem. Thanks for your detailed review though. I am working on addressing the other suggested changes.

nordic-krch commented on subsys/logging/log_core.c:
```diff
@@ -352,6 +359,42 @@ void log_n(const char *str,
 	}
 }
 
+int log_format_set(char *backend_name, uint32_t log_type)
```
should be moved to `log_backend.h`

aasthagr replied:
Addressed.

nordic-krch reviewed (changes requested):
Nice feature. I left few comments.

aasthagr committed:
logging: Add API to get log_backend instance by name.

Added helper API to the logging backend interface to get the
log backend instance by name.

Signed-off-by: Aastha Grover <aastha.grover@intel.com>

aasthagr committed:
logging: Add API's to switch logging formats at runtime.

Add support for new API's log_backend_format_set and
log_format_set_all_active_backends to switch the logging format
for one backend and for all the active backends respectively.
Using format_set function pointer in log_backend_api struct as a hook
to set the log format at runtime in the backends. Add function pointer
table with helper functions to be used as a way to select log format.
Add supported log format types as macros to use with the API.

Signed-off-by: Aastha Grover <aastha.grover@intel.com>

aasthagr commented on include/logging/log_backend.h:
```diff
@@ -254,6 +254,15 @@ static inline const struct log_backend *log_backend_get(uint32_t idx)
+ *
+ * @retval Pointer to the backend instance if found, NULL if backend is not found.
+ */
+extern const struct log_backend *log_backend_get_by_name(const char *backend_name);
```
Addressed.

aasthagr commented on include/logging/log_core.h:
```diff
@@ -895,6 +896,30 @@ do {									       \
+ * @retval -EINVAL  If the input is invalid.
+ * @retval	0 for success.
+ */
+int log_format_set(char *backend_name, uint32_t log_type);
```
Addressed.

aasthagr commented on include/logging/log_core.h:
```diff
@@ -895,6 +896,30 @@ do {									       \
+ *
+ * @retval Pointer to the last backend that failed, NULL for success.
+ */
+const struct log_backend *log_format_set_all_active_backends(uint32_t log_type);
```
Addressed.

aasthagr commented on include/logging/log_core.h:
```diff
@@ -895,6 +896,30 @@ do {									       \
+/** @brief Function pointer table for log output formats.
+ *
+ */
+extern void (*log_format_table[])(const struct log_output*, struct log_msg2*, uint32_t);
```
@nordic-krch Made the suggested changes. Please re-review and let me know.

aasthagr commented on include/logging/log_core.h:
```diff
@@ -14,6 +14,7 @@
 #include <stdarg.h>
 #include <syscall.h>
 #include <sys/util.h>
+#include <logging/log_output.h>
```
Addressed.

aasthagr commented on subsys/logging/Kconfig.backends:
```diff
@@ -33,7 +33,7 @@ config LOG_BACKEND_UART_OUTPUT_TEXT
 	  Output in text.
 
-config LOG_BACKEND_UART_SYST_ENABLE
+config LOG_BACKEND_UART_OUTPUT_SYST
```
@MaureenHelm Since I have defined LOG_BACKEND_UART_OUTPUT_DEFAULT which gives us the option to select from LOG_BACKEND_UART_OUTPUT_TEXT, LOG_BACKEND_UART_OUTPUT_SYST or LOG_BACKEND_UART_OUTPUT_DICTIONARY. Just wanted to be consistent with the naming of formats for backends here. I have a plan to do so for every backend. Also, this wont go away with v1 logging because from my understanding we still need the functionality that each backend can have a different logging format when multiple backends are active. Let me know if this still doesn't make sense.

MaureenHelm replied:
We can simplify this by having one default output Kconfig for all backends:

```
choice LOG_BACKEND_OUTPUT_DEFAULT
        prompt "Default output format for all backends"

config LOG_BACKEND_OUTPUT_DEFAULT_TEXT
        bool "Text"

config LOG_BACKEND_OUTPUT_DEFAULT_SYST
        bool "Sys-T"

config LOG_BACKEND_OUTPUT_DEFAULT_DICT
        bool "Dictionary"

endchoice
```

Or, if we need to specify a different default for each backend, then use a Kconfig template to generate them instead of hand-coding all the possible combinations. See for example:
- `subsys/logging/Kconfig.template.log_config`
- `subsys/net/Kconfig.template.log_config.net`
- `subsys/usb/class/Kconfig.template.composite_device_number`

aasthagr replied:
@MaureenHelm That was a very good suggestion. I have leveraged the Kconfig template to add configurations and added support for all the backends. Please re-review.

aasthagr committed:
logging: Add support to backends to switch log format at runtime.

Adding mechanism to all the backends to switch the logging formats
at runtime while leveraging the function pointer table based upon
the Kconfigs which also cleans up the logging backend design.
Also demonstrate the working API with the changes
to syst sample. Clean up Kconfig for backends and add a standard
template to generate the Kconfigs for logging formats to be used
by all backends.

Signed-off-by: Aastha Grover <aastha.grover@intel.com>

aasthagr commented on subsys/logging/log_core.c:
```diff
@@ -352,6 +369,24 @@ void log_n(const char *str,
+const struct log_backend *log_format_set_all_active_backends(uint32_t log_type)
+{
+	const struct log_backend *backend;
+	const struct log_backend *failedBackend = NULL;
```
Renamed.

aasthagr commented on subsys/logging/log_core.c:
```diff
@@ -58,6 +59,22 @@ LOG_MODULE_REGISTER(log);
+log_format_func_t format_table[] = {
+	log_output_msg2_process,
+	IS_ENABLED(CONFIG_LOG_MIPI_SYST_ENABLE) ? log_output_msg2_syst_process : NULL,
+	IS_ENABLED(CONFIG_LOG_DICTIONARY_SUPPORT) ? log_dict_output_msg2_process : NULL
```
Addressed as suggested. Thanks!!

nordic-krch commented on subsys/logging/log_core.c:
```diff
@@ -58,6 +59,24 @@ LOG_MODULE_REGISTER(log);
+
+log_format_func_t log_format_func_t_get(uint32_t log_type)
+{
+	return format_table[log_type];
```
should check if log_type is within array.

aasthagr replied:
@nordic-krch This check is already in place in "log_backend_format_set" API.

nordic-krch commented on subsys/logging/log_core.c:
```diff
@@ -58,6 +59,24 @@ LOG_MODULE_REGISTER(log);
+	return format_table[log_type];
+}
+
+int log_format_table_size(void)
```
size_t

aasthagr replied:
Changed the type to size_t.

nordic-krch commented on include/logging/log_output.h:
```diff
@@ -50,6 +50,15 @@ extern "C" {
  */
 #define LOG_OUTPUT_FLAG_FORMAT_SYST		BIT(7)
```
shouldn't this flag be deprecated?

aasthagr replied:
I think this flag is still being used in parts of v1 logging routines (put, sync_string, sync_hexdump) in the backends.

nordic-krch replied:
ok, can you add comment  that it is deprecated and can be used only when `CONFIG_LOG1` is enabled. It will be easier to find it when v1 will be removed after 3.1.

aasthagr replied:
Done.

aasthagr replied:
@MaureenHelm Using __DEPRECATED_MACRO for LOG_OUTPUT_FLAG_FORMAT_SYST results in warnings which cause failing testcases in twister run.

nordic-krch commented on subsys/logging/Kconfig.template.log_format_config:
```diff
@@ -0,0 +1,52 @@
+	depends on LOG2
+	select LOG_DICTIONARY_SUPPORT
+	help
+	  Backend is in dictionary-based logging output mode. Supported only for UART backend.
```
"Supported only for UART backend?" still true?

aasthagr replied:
@nordic-krch Prior this implementation, this flag only existed for UART backend. Also the support for Dictionary Format as well was only in UART backend. This Kconfig template would enable all other backends as well for dictionary format, although I wasn't sure if the support can be validated therefore provided this helper info just incase. Thoughts?

nordic-krch commented on subsys/logging/Kconfig.template.log_format_config:
```diff
@@ -0,0 +1,52 @@
+	depends on LOG2
+	select LOG_BACKEND_$(backend)_OUTPUT_DICTIONARY
+	help
+	  Dictionary-based logging output in binary. Supported only for UART backend.
```
This one seems be supported for all backend

nordic-krch commented on subsys/logging/Kconfig.template.log_format_config:
```diff
@@ -0,0 +1,52 @@
+	depends on LOG2
+	select LOG_BACKEND_$(backend)_OUTPUT_DICTIONARY
+	help
+	  Dictionary-based logging output in hexadecimal. Supported only for UART backend.
```
this one is for UART only. Maybe it shouldn't be in template but only added in uart configuration.

aasthagr replied:
@nordic-krch Moved this to uart configuration. Thanks.

nordic-krch reviewed (dismissed):
looks good

dcpleung commented on subsys/logging/Kconfig.backends:
```diff
@@ -17,44 +17,16 @@ config LOG_BACKEND_UART_ASYNC
-	bool "Dictionary (binary)"
-	depends on LOG2
-	select LOG_BACKEND_UART_OUTPUT_DICTIONARY
+	select LOG_BACKEND_$(backend)_OUTPUT_DICTIONARY
 	help
-	  Dictionary-based logging output in binary.
-
-endchoice
+	  Dictionary-based logging output in hexadecimal. Supported only for UART backend.
```
This presents an option `LOG_BACKEND_UART_OUTPUT_DICTIONARY_HEX` in addition to the output choice block in the template. Note that `..._BIN` is enclosed in a `choice`/`endchoice` pair, but `..._HEX` is outside the choice block. So even if `..._BIN` is enabled (i.e. selected in the choice block), `..._HEX` can be set independently. But in reality, these two options are mutually exclusive.

aasthagr replied:
@dcpleung Corrected this. Thanks for noticing this.

dcpleung commented on subsys/logging/Kconfig.backends:
```diff
@@ -319,30 +326,18 @@ config LOG_BACKEND_FS
 
 config LOG_BACKEND_FS_OUTPUT_DICTIONARY_BIN
 	bool "Dictionary (binary)"
-	depends on LOG2
-	select LOG_BACKEND_FS_OUTPUT_DICTIONARY
 	help
 	  Dictionary-based logging output in binary.
 
-endchoice
+endif #LOG_BACKEND_FS_OUTPUT_DICTIONARY
```
There is no need for this, as this kconfig is not being used by the backend.

aasthagr replied:
@dcpleung The option "LOG_BACKEND_FS_OUTPUT_DICTIONARY_BIN"  does exist for the FS backend. Not sure what you meant here?

dcpleung replied:
This is not being used in the backend code.

aasthagr replied:
True. Removed this option from FS backend Kconfig.

dcpleung commented on subsys/logging/Kconfig.template.log_format_config:
```diff
@@ -0,0 +1,38 @@
+
+config LOG_BACKEND_$(backend)_OUTPUT_SYST
+	bool "MIPI SyS-T"
+	depends on LOG_BACKEND_$(backend)
```
No need for this, as the if in `Kconfig.backends` already implies this.

aasthagr replied:
Okay. Removed this dependency.

nashif commented on include/logging/log_backend.h:
```diff
@@ -48,6 +48,8 @@ struct log_backend_api {
 	void (*dropped)(const struct log_backend *const backend, uint32_t cnt);
 	void (*panic)(const struct log_backend *const backend);
 	void (*init)(const struct log_backend *const backend);
+	int (*format_set)(const struct log_backend *const backend,
```
why size_t? This just an integer indicating the type.

aasthagr replied:
@nashif log_format_table_size() was changed to return type "size_t" as I think it is a general good practice returning a size should use size_t (when the ARRAY_SIZE operator is involved) . 

Since there were comparisons in the API log_backend_format_set [ log_type >= log_format_table_size() ] and also we do not want a negative value for log_type therefore considered size_t type for log_type. Let me know if this doesn't make sense?

nashif replied:
> @nashif log_format_table_size() was changed to return type "size_t" as I think it is a general good practice returning a size should use size_t (when the ARRAY_SIZE operator is involved) .

Using it for log_format_table_size is fine, I am wondering if using it here...
```
static size_t log_format_current = CONFIG_LOG_BACKEND_ADSP_OUTPUT_DEFAULT;
```

and elsewhere is the right choice of type, although it makes sense to use size_t for array indexes, but this is not clear in the above code for example.

aasthagr replied:
I have changed the size_t type to uint32_t in all the places except for log_format_table_size where I still use "size_t".

nashif commented:
Also, this PR is missing a test or sample code that demonstrates runtime switching of formats.. Please provide tests/show how this API is to be used with a sample

nashif commented on samples/subsys/logging/syst/src/main.c:
```diff
@@ -58,6 +60,10 @@ void main(void)
 
+	size_t log_type = LOG_OUTPUT_TEXT;
+
+	log_backend_format_set(log_backend_get_by_name("log_backend_uart"), log_type);
```
I would put all existing log messages in another function and then use first format )default) to display the, switch, display them again in the same format, switch back and display the original format, this way we do not lose functionality of the sample originally designed for demonstrating sys-t

aasthagr replied:
@nashif Sure. Have updated the main.c for sys-t sample as suggested. Please re-review.

MaureenHelm reviewed (commented):
Have you tested any of the other backends besides UART?

MaureenHelm commented:
> Have you tested any of the other backends besides UART?

With the following change, the sample works with native_posix:
```diff
diff --git a/samples/subsys/logging/syst/src/main.c b/samples/subsys/logging/syst/src/main.c
index c0ae6a4ac6..7bb9768d1e 100644
--- a/samples/subsys/logging/syst/src/main.c
+++ b/samples/subsys/logging/syst/src/main.c
@@ -115,12 +115,12 @@ void main(void)
 
        uint32_t log_type = LOG_OUTPUT_TEXT;
 
-       log_backend_format_set(log_backend_get_by_name("log_backend_uart"), log_type);
+       log_backend_format_set(log_backend_get_by_name("log_backend_native_posix"), log_type);
 
        log_msgs();
 
        log_type = LOG_OUTPUT_SYST;
-       log_backend_format_set(log_backend_get_by_name("log_backend_uart"), log_type);
+       log_backend_format_set(log_backend_get_by_name("log_backend_native_posix"), log_type);
 
        log_msgs();
 #endif
```

nashif commented:
@aasthagr  please follow the commit message guidelines https://docs.zephyrproject.org/latest/contribute/index.html#commit-guidelines, i.e. subsystem is logging and you and you go into too much details about things that are in the diff, you should have the change description with your logic or reasoning for the changes instead of code walkthrough.

aasthagr commented:
> Have you tested any of the other backends besides UART?



> @aasthagr please follow the commit message guidelines https://docs.zephyrproject.org/latest/contribute/index.html#commit-guidelines, i.e. subsystem is logging and you and you go into too much details about things that are in the diff, you should have the change description with your logic or reasoning for the changes instead of code walkthrough.

Okay. I agree the message was quite cluttered. I have changed the commit messages. Please re-review and let me know if any other changes are required.

nashif commented on include/logging/log_ctrl.h:
```diff
@@ -170,6 +170,15 @@ void log_backend_enable(struct log_backend const *const backend,
+ *
+ * @retval Pointer to the backend instance if found, NULL if backend is not found.
+ */
+extern const struct log_backend *log_backend_get_by_name(const char *backend_name);
```
why do we need the extern here?

aasthagr replied:
I think we do not need it as such. I can remove the extern.
